src/connection.cpp

```diff
@@ -41,295 +41,306 @@ namespace oxen::quic
+    int connection_callbacks::on_early_data_rejected(ngtcp2_conn* _conn [[maybe_unused]], void* user_data)
+    {
+        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
 
-            auto& conn = *static_cast<Connection*>(user_data);
-            assert(_conn == conn);
+        auto& conn = *static_cast<Connection*>(user_data);
+        assert(_conn == conn);
 
-            (void)conn;
-            (void)_conn;
+        log::info(log_cat, "Client resetting early streams; 0-rtt rejected by server!");
+        conn.revert_early_streams();
 
-            return 0;
-        }
-    };
+        return 0;
+    }
 
-    void Connection::set_close_quietly()
+    int connection_callbacks::recv_stateless_reset(
+            ngtcp2_conn* _conn [[maybe_unused]], const ngtcp2_pkt_stateless_reset* sr, void* user_data)
     {
         log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
 
-        _close_quietly = true;
+        auto& conn = *static_cast<Connection*>(user_data);
+        assert(_conn == conn);
+
+        return conn.recv_stateless_reset(gtls_reset_token::make_copy(sr->stateless_reset_token, sr->rand));
     }
 
-    void Connection::set_new_path(Path new_path)
+    int Connection::recv_stateless_reset(std::shared_ptr<gtls_reset_token> tok)
     {
-        _endpoint.call([this, new_path]() { _path = new_path; });
+        log::trace(log_cat, "Client recv_stateless_reset cb called...");
+
+        if (auto it = _endpoint.reset_token_map.find(tok); it != _endpoint.reset_token_map.end())
+        {
+            log::warning(
+                    log_cat,
+                    "Received stateless reset for connection ({}) on path: {}; closing immediately!",
+                    _ref_id,
+                    _path);
+            _endpoint.drop_connection(*this, io_error{CONN_STATELESS_RESET});
+        }
+        else
+            log::debug(log_cat, "Could not match received stateless reset to any connection!");
+        return 0;
     }
 
-    int Connection::recv_token(const uint8_t* token, size_t tokenlen)
+    int Connection::client_path_validation(const ngtcp2_path* path, bool success, uint32_t flags)
     {
-        // This should only be called by the client, and therefore this will always have a value
-        assert(not remote_pubkey.empty());
-        _endpoint.store_path_validation_token(remote_pubkey, {token, tokenlen});
+        log::trace(log_cat, "Client path_validation cb called...");
+        assert(is_outbound());
+
+        if (flags & NGTCP2_PATH_VALIDATION_FLAG_PREFERRED_ADDR)
+        {
+            set_remote_addr(path->remote);
+            log::debug(log_cat, "Client set new remote ({}) on successful path validation...", _path.remote);
+        }
+        else
+            log::warning(
+                    log_cat,
+                    "Client path validation {}; no address was provided by server...",
+                    success ? "succeeded" : "failed");
```

Comment:
Is this supposed to be a *warning* when it prints "Client path validation succeeded"?